### PR TITLE
lake

### PR DESCRIPTION
**Pull request by playerInEminor**
Yifei Fang, Sabrina Araujo

### CODE REVIEW

File: 01-python/outdoors/lake.py [hunk 1]
```diff
@@ -1,3 +1,15 @@
...
+
+def draw_lake(area=1, shape="square"):
+    if(shape=="square"):
+        for i in range(math.sqrt(area)):
+            print("-")
+        for i in range(math.sqrt(area)):
+            print("|" + math.sqrt(area)*" " + "|")
+        for i in range(math.sqrt(area)):
+            print("-")
+    # n =math.sqrt(area) - 1
```
Comment (gallettilance) [inline 01-python/outdoors/lake.py]: please remove comments from the code

**Comment by playerInEminor** [comment]
Sorry, but I don't undestand how can I squash these commits as I have push them to github.

**Comment by gallettilance** [comment]
Try to follow [these instructions](https://github.com/wprig/wprig/wiki/How-to-squash-commits) or come to office hours

**Comment by gallettilance** [comment]
Oh I see the problem. You are working on the master branch. In this case you need to create a new remote called `upstream` pointing to `https://github.com/gallettilance/CS506-Spring2020.git` (run:
```bash
git remote add upstream https://github.com/gallettilance/CS506-Spring2020.git
```
and then rebase on `upstream/master` instead of just `master` after fetching the latest upstream changes.
i.e.
```bash
git fetch upstream
git rebase -i upstream/master`
```

File: 01-python/outdoors/lake.py [hunk 2]
```diff
@@ -1,3 +1,34 @@
-def draw_lake():
-    print("lake not found")
+import math
+
+
+def draw_lake(area=1, shape="rectangle", width=1, length=1):
+    """
+    Print a lake
+
+    The shape of lake should be square or rectangle.
```
Comment (gallettilance) [inline 01-python/outdoors/lake.py]: You should list the parameters as in [this example](https://www.geeksforgeeks.org/python-docstrings/)

File: 01-python/outdoors/lake.py [hunk 3]
```diff
@@ -1,3 +1,34 @@
...
+    If the lake is square, the area should be a square number.
+    If the lake is rectangle, the given width and length should compatible with the area.
+    """
+    if shape == "square":
+        n = math.sqrt(area)
```
Comment (gallettilance) [inline 01-python/outdoors/lake.py]: you don't need to import the `sqrt` function to check if a number is a square :)

File: 01-python/outdoors/lake.py [hunk 4]
```diff
@@ -1,3 +1,34 @@
...
+    if shape == "rectangle":
+        if width*length != area:
+            print("width and length are wrong!")
+        else:
+            print(" " + length * "-")
+            for i in range(width):
+                print("|" + length * " " + "|")
+            print(" " + length * "-")
+            return
+    print("Sorry, now the shape should be square or rectangle.")
```
Comment (gallettilance) [inline 01-python/outdoors/lake.py]: You should raise an error instead of printing something.

Comment (gallettilance) [reply 01-python/outdoors/lake.py]: and you should raise this error as soon as you can raise it.

File: 01-python/infrastructure/road.py [hunk 5]
```diff
@@ -1,3 +1,8 @@
-def draw_road():
-    print("road not found")
+def draw_road(length=10, axis="horizontal"):
```
Comment (gallettilance) [inline 01-python/infrastructure/road.py]: please remove the changes to this file

File: 01-python/outdoors/lake.py (hunk 3 above)
Comment (playerInEminor) [reply 01-python/outdoors/lake.py]: Sorry, I don't understand how can I check whether a number is a square without `sqrt`.

Comment (gallettilance) [reply 01-python/outdoors/lake.py]: A very short google: https://www.geeksforgeeks.org/check-if-a-number-is-perfect-square-without-finding-square-root/

File: 01-python/infrastructure/road.py [hunk 6]
```diff
@@ -1,3 +1,2 @@
 def draw_road():
     print("road not found")
-    return
```
Comment (gallettilance) [inline 01-python/infrastructure/road.py]: Please revert this change

File: 01-python/outdoors/lake.py [hunk 7]
```diff
@@ -1,3 +1,46 @@
...
+
+    The shape of lake should be square or rectangle.
+    If the lake is square, the area should be a square number.
+    If the lake is rectangle, the given width and length should compatible with the area.
+    """
+    if shape == "square":
+        n = is_perfect_square(area)
+        if not n:
+            raise Exception("area of square should be square number!")
+            return
```
Comment (gallettilance) [inline 01-python/outdoors/lake.py]: raising the exception will return
```suggestion

```

File: 01-python/outdoors/lake.py [hunk 8]
```diff
@@ -1,3 +1,46 @@
...
+            print(" " + length * "-")
+            for i in range(width):
+                print("|" + length * " " + "|")
+            print(" " + length * "-")
+            return
+    raise Exception("Sorry, now the shape should be square or rectangle.")
     return
+
+
+def is_perfect_square(n):
```
Comment (gallettilance) [inline 01-python/outdoors/lake.py]: the name of this function implies that it returns a boolean. Having different return types is strange and is difficult to surface as expected side effects to the caller. Here is a different/cleaner way of doing this using errors:

```python
...
    if shape == "square":
        n = get_perfect_square_root(area) #this will raise an error if area is not a perfect square
        for i in range(n):
            print("|" + n*" " + "|")
        print(" " + n * "-")
        return
...

def get_perfect_square_root(n):
    i = 1
    while i * i <= n:
        if (n % i == 0) and (n / i == i):
            return i
        i = i + 1
    raise ValueError("input is not a perfect square")
```

Comment (gallettilance) [reply 01-python/outdoors/lake.py]: optionally, if you would like to raise a different error in the caller function, you can use
```python
try:
    i = get_perfect_square_root(area)
catch ValueError as e:
# do something with the error or raise a different error
